src/polycap-description.c

```diff
@@ -0,0 +1,434 @@
+	printf("Type in the amount of threads to use (max %d):\n",thread_max);
+	scanf("%d",&thread_cnt);
+	printf("%d threads out of %d selected.\n",thread_cnt, thread_max);
+
+	// Prepare arrays to save results
+	sum_weights = malloc(sizeof(double)*n_energies);
+	if(sum_weights == NULL){
+		printf("Could not allocate sum_weights memory.\n");
+		exit(1);
+	}
+	for(i=0; i<n_energies; i++) sum_weights[i] = 0.;
+printf("Here1\n");
+printf("%f keV - %f keV\n",energies[0],energies[n_energies-1]);
+//OpenMP loop
+#pragma omp parallel \
+	default(shared) \
+	private(i, j) \
+	firstprivate(description, n_energies, energies) \
+	num_threads(thread_cnt)
+{
+	int thread_id = omp_get_thread_num();
+	polycap_rng *rng;
+	unsigned int seed;
+	polycap_vector3 start_coords, start_direction, start_electric_vector;
+	double r; //random number
+	double n_shells; //amount of capillary shells in polycapillary
+	int boundary_check;
+	double src_rad_x, src_rad_y, phi; //distance from source centre in x and y direction and angle phi from x axis
+	double src_start_x, src_start_y;
+	double pc_rad, pc_x, pc_y; //pc radius and coordinates to direct photon to
+	polycap_photon *photon;
+	int iesc=-1, k;
+	int64_t istart=0; //amount of started photons
+	double *weights;
+
+	weights = malloc(sizeof(double)*n_energies);
+	if(weights == NULL){
+		printf("Could not allocate weights memory.\n");
+		exit(1);
+	}
+	for(k=0; k<n_energies; k++) weights[k] = 0.;
+
+	// Create new rng
+#ifdef _WIN32
+	rand_s(&seed);
+#else
+	FILE *random_device;
+	if((random_device = fopen("/dev/urandom", "r")) == NULL){
+		printf("Could not open /dev/urandom\n");
+		exit(2);
+	}
+	fread(&seed, sizeof(unsigned long int), 1, random_device);
+	fclose(random_device);
+#endif
+	rng = polycap_rng_new(seed);
+
+printf("%f keV - %f keV\n",energies[0],energies[n_energies-1]);
+printf("Here2, Thread%d\n",thread_id);
+
+	#pragma omp for nowait
+	for(j=0; j < icount; j++){
+		do{
+			// Obtain photon start coordinates
+			n_shells = round(sqrt(12. * description->n_cap - 3.)/6.-0.5);
+			if(n_shells == 0.){ //monocapillary case
+				r = polycap_rng_uniform(rng);
+				start_coords.x = (2.*r-1) * description->profile->cap[0];
+				r = polycap_rng_uniform(rng);
+				start_coords.y = (2.*r-1) * description->profile->cap[0];
+				start_coords.z = 0.;
+			} else { // polycapillary case
+				// select random coordinates, check whether they are inside polycap boundary
+				boundary_check = -1;
+				do{
+					r = polycap_rng_uniform(rng);
+					start_coords.x = (2.*r-1) * description->profile->ext[0];
+					r = polycap_rng_uniform(rng);
+					start_coords.y = (2.*r-1) * description->profile->ext[0];
+					start_coords.z = 0.;
+					boundary_check = polycap_photon_within_pc_boundary(description->profile->ext[0], start_coords);
+				} while(boundary_check == -1);
+			}
+
+printf("Here3\n");
+			// Obtain point from source as photon origin, determining photon start_direction
+			r = polycap_rng_uniform(rng);
```

Comment:
Eigenlijk zouden de source parameters uit de `polycap_description` moeten gehaald worden en in hun eigen `polycap_source` moet gezet worden. Eigenlijk maakt dat geen deel uit van de beschrijving van het polycapillair, en worden toch enkel gebruikt in `polycap_description_get_transmission_efficiencies`, welke niet zal gebruikt worden in XMI-MSIM, en waardoor je me dus zal vragen een hoop nutteloze source parameters mee te geven wanneer ik `polycap_description_new` zal oproepen.

Ik weet dat die source parameters deel uitmaken van 1 van Laszlo's input-files... Ik denk dat deze source specifieke parameters in dat geval best naar hun eigen bestand worden verhuisd, dat dan te vermelden wordt in de main input-file.

Wat denk je hiervan?

Reply:
Zulks een `polycap_source` zou dus een argument worden van `polycap_description_get_transmission_efficiencies`

Reply:
Geen probleem mee, en inderdaad nog iets netter :)

Reply:
en `polycap_description_get_transmission_efficiencies` wordt mss wel gebruikt voor het focussing polycapilair in MSIM he, maar kbegrijp je punt :)
In een van mijn laatste versies van polycap, voor ik het nu met je API begon te herschrijven, was ik immers ook al begonnen met die zaken uit elkaar te halen met aparte structures. Enkel was de code toen nog zo'n nest dat alle structures in alle functies werden meegegeven ongeveer :P